### PR TITLE
Fix smartSummarize to properly handle alignTo start time and intervals smaller than data StepTime

### PR DESCRIPTION
**Pull request by carrieedwards**
This PR rewrites the smartSummarize function in order to correct two issues that were discovered (similar to the recent fix to hitcount, see #132).  

When the interval parameter was set to a value smaller than the fetched data's step between data points, it results were incorrect. The results were a copy of the data points for each series in the list.  For example:
smartSummarize(metric1,'6m','sum', 'minutes')
[]*types.MetricData{{"metric1", 0, 1}: {types.MakeMetricData("metric1", []float64{2, 4, 6}, 600, 0)},} // 10m step

The result will be []float64{2, 4, 6}. Issuing the same query with the same data in Graphite-web yielded results of [2, 4, None, 6, None]. This is due to the interval being smaller than the step, and therefore some resulting values will be NaN.

Additionally, in Graphite-web's implement of smartSummarize,  the start time of the request is adjusted based on the  alignTo value, and the data is re-fetched. In CarbonAPI, the start time was adjusted, but data was not re-fetched.
Major changes in this PR:

The smartSummarize function was completely rewritten to match the behavior of Graphite web, as well as adding some optimizations in the algorithm. This helps address the issue with intervals smaller than the data's step, as well as fixes discrepancies in results that were occurring with CarbonAPI tests that were tested in Graphite web.
The Metrics() function in pkg/parser.go was updated to adjust the start time of a metric request with a target of smartSummarize, so that the start time reflects the alignTo parameter's value. This is intended to prevent needing to re-fetch the data during smartSummarize function execution as is done in Graphite web.

Additionally, more testcases were added, including ones that cover the situation where the specific interval is smaller than the fetched data's step, as well as test cases translated from Graphite web for smartSummarize.

### CODE REVIEW

File: expr/functions/smartSummarize/function_test.go [hunk 1]
```diff
@@ -130,6 +131,521 @@ func TestEvalSummarize(t *testing.T) {
...
+			14400,
+		},
+	}
+
+	for _, tt := range tests {
+		th.TestSummarizeEvalExpr(t, &tt)
+	}
+}
+
+func TestSmartSummarizeAlignToWeeksThursday(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (carrieedwards) [reply expr/functions/smartSummarize/function_test.go:336]: Yeah, for now I just used 'weeks' because IntervalString fails to parse weeks4. But I will work on fixing that in IntervalString

File: expr/functions/smartSummarize/function_test.go [hunk 2]
```diff
@@ -130,6 +131,521 @@ func TestEvalSummarize(t *testing.T) {
...
+			240,
+		},
+	}
+
+	for _, tt := range tests {
+		th.TestSummarizeEvalExpr(t, &tt)
+	}
+}
+
+func TestSmartSummarizeAlignTo1Year(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (carrieedwards) [reply expr/functions/smartSummarize/function_test.go:208]: Working on adding support the other interval formats as well

File: pkg/parser/parser.go [hunk 3]
```diff
@@ -861,3 +881,47 @@ func parseString(s string) (string, string, error) {
...
 	return s[:i], s[i+1:], nil
 }
+
+func StartAlignTo(start int64, alignTo string) (int64, error) {
+	var newDate time.Time
+
+	if IsDigit(alignTo[0]) {
+		re := regexp.MustCompile(`^[0-9]+`)
+		alignTo = re.ReplaceAllString(alignTo, "")
+	}
```
Comment (npazosmendez) [inline pkg/parser/parser.go:891]: nit: this panics if `len(alignTo) == 0`, I think we can just do the replace without checking:
```suggestion
	re := regexp.MustCompile(`^[0-9]+`)
	alignTo = re.ReplaceAllString(alignTo, "")
```

File: expr/functions/smartSummarize/function.go [hunk 4]
```diff
@@ -51,6 +64,9 @@ func (f *smartSummarize) Do(ctx context.Context, e parser.Expr, from, until int6
 	if err != nil {
 		return nil, err
 	}
+	if bucketSizeInt32 == 0 {
```
Comment (npazosmendez) [inline expr/functions/smartSummarize/function.go:67]: This is still not checking for negative buckets, which cause panics. We should add a test passing a negative bucket and assert it errors but does not panic

**Review by npazosmendez — APPROVED** [review]
LGTM

For the alignment tests: I think we should change the `TestSummarizeEvalExpr` helper so the from/to times can be adjusted; otherwise the alignment is not being tested. But I don't want to put that as a blocker now

**Comment by carrieedwards** [comment]
> LGTM
> 
> For the alignment tests: I think we should change the `TestSummarizeEvalExpr` helper so the from/to times can be adjusted; otherwise the alignment is not being tested. But I don't want to put that as a blocker now

I will do that and update the tests in a separate PR, since this PR is already a pretty large one.
